File: src/claudebox_daemon/handlers/boards.py
"""Board CRUD — workspace-scoped HTTP adapters for board management."""

from fastapi import APIRouter
from fastapi.responses import PlainTextResponse

from ._models import (
    AssignTicketsRequest,
    CreateSwimlaneRequest,
    MoveTicketRequest,
    RenameBoardRequest,
    RenameStateRequest,
    RenameSwimlaneRequest,
    ReorderStatesRequest,
    ReorderSwimlanesRequest,
)
from ._shared import WorkspaceDep


router = APIRouter(prefix="/api/workspaces/{workspace_id}")


# Board CRUD
# ----------------------------------------------------------------------------------------------


@router.get("/boards")
async def list_boards(svc: WorkspaceDep):
    """List discovered boards in the workspace."""

    return {"boards": svc.board_service.list_all()}


@router.get("/boards/{board_id}")
async def get_board(svc: WorkspaceDep, board_id: str):
    """Get full board state with resolved ticket titles and session status."""

    return svc.board_service.get(board_id)


@router.patch("/boards/{board_id}")
async def rename_board(svc: WorkspaceDep, board_id: str, body: RenameBoardRequest):
    """Rename a board — set the name: field in its board.yaml."""

    return svc.board_service.rename(board_id, body.name)


# Tickets
# ----------------------------------------------------------------------------------------------


@router.get("/boards/{board_id}/tickets/{ticket_path:path}/content")
async def get_ticket_content(svc: WorkspaceDep, board_id: str, ticket_path: str):
    """Return raw markdown content of a ticket file."""

    content = svc.board_service.read_ticket_content(board_id, ticket_path)
    return PlainTextResponse(content, media_type="text/markdown")


@router.patch("/boards/{board_id}/tickets/{ticket_path:path}/move")
async def move_ticket(svc: WorkspaceDep, board_id: str, ticket_path: str, body: MoveTicketRequest):
    """Move a ticket between columns and/or swimlanes, optionally at a specific index."""

    return svc.board_service.move(
        board_id,
        ticket_path,
        column=body.column,
        swimlane=body.swimlane,
        index=body.index,
    )


@router.delete("/boards/{board_id}/tickets/{ticket_path:path}")
async def archive_ticket(svc: WorkspaceDep, board_id: str, ticket_path: str):
    """Archive a ticket — remove from YAML, file stays on disk."""

    svc.board_service.archive(board_id, ticket_path)
    return {"status": "archived"}


@router.post("/boards/{board_id}/assign")
async def assign_tickets(svc: WorkspaceDep, board_id: str, body: AssignTicketsRequest):
    """Batch assign tickets to new sessions."""

    sessions = await svc.board_service.assign(
        board_id,
        body.tickets,
        parallel=body.parallel,
    )
    return {"sessions": sessions}


# Swimlanes
# ----------------------------------------------------------------------------------------------


@router.post("/boards/{board_id}/swimlanes")
async def create_swimlane(svc: WorkspaceDep, board_id: str, body: CreateSwimlaneRequest):
    """Create a new swimlane."""

    return svc.board_service.create_swimlane(board_id, body.name)


@router.patch("/boards/{board_id}/swimlanes/{swimlane_id}")
async def rename_swimlane(
    svc: WorkspaceDep,
    board_id: str,
    swimlane_id: str,
    body: RenameSwimlaneRequest,
):
    """Rename an existing swimlane."""

    return svc.board_service.update_swimlane(board_id, swimlane_id, body.name)


@router.delete("/boards/{board_id}/swimlanes/{swimlane_id}")
async def delete_swimlane(svc: WorkspaceDep, board_id: str, swimlane_id: str):
    """Delete a swimlane. Tickets in it become unsorted."""

    svc.board_service.remove_swimlane(board_id, swimlane_id)
    return {"status": "deleted"}


@router.patch("/boards/{board_id}/swimlanes/reorder")
async def reorder_swimlanes(svc: WorkspaceDep, board_id: str, body: ReorderSwimlanesRequest):
    """Reorder swimlanes to match the given ID order."""

    return {"swimlanes": svc.board_service.reorder(board_id, body.ids)}


# States
# ----------------------------------------------------------------------------------------------


@router.patch("/boards/{board_id}/states/reorder")
async def reorder_states(svc: WorkspaceDep, board_id: str, body: ReorderStatesRequest):
    """Reorder columns/states to match the given key order."""

    return {"states": svc.board_service.reorder_columns(board_id, body.keys)}


@router.patch("/boards/{board_id}/states/{state_id}")
async def rename_state(svc: WorkspaceDep, board_id: str, state_id: str, body: RenameStateRequest):
    """Rename a state's display label. Folder and ID stay unchanged."""

    return await svc.board_service.rename_state(board_id, state_id, body.label)

File: src/claudebox_frontend/src/api/boards.js
/** Board API client — workspace-scoped board CRUD operations. */

import { workspaceFetch } from './apiClient'

/** List all discovered boards in the workspace. */
export async function listBoards() {
  const res = await workspaceFetch('/boards')
  if (!res.ok) {
    throw new Error('Failed to fetch boards')
  }
  return res.json()
}

/** Get full board state with resolved ticket titles. */
export async function getBoard(boardId) {
  const res = await workspaceFetch(`/boards/${boardId}`)
  if (!res.ok) {
    throw new Error('Failed to fetch board')
  }
  return res.json()
}

/** Rename a board by setting its name: field. */
export async function renameBoard(boardId, name) {
  const res = await workspaceFetch(`/boards/${boardId}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  })
  if (!res.ok) {
    throw new Error('Failed to rename board')
  }
  return res.json()
}

/** Get raw markdown content of a ticket file. */
export async function getTicketContent(boardId, ticketPath) {
  const res = await workspaceFetch(
    `/boards/${boardId}/tickets/${encodeURIComponent(ticketPath)}/content`,
  )
  if (!res.ok) {
    throw new Error('Failed to fetch ticket content')
  }
  return res.text()
}

/** Move a ticket between columns and/or swimlanes, optionally at a specific index. */
export async function moveTicket(boardId, ticketPath, { column, swimlane, index } = {}) {
  const body = { column, swimlane }
  if (index !== undefined) {
    body.index = index
  }
  const res = await workspaceFetch(
    `/boards/${boardId}/tickets/${encodeURIComponent(ticketPath)}/move`,
    {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    },
  )
  if (!res.ok) {
    throw new Error('Failed to move ticket')
  }
  return res.json()
}

/** Archive a ticket — remove from YAML, file stays on disk. */
export async function archiveTicket(boardId, ticketPath) {
  const res = await workspaceFetch(`/boards/${boardId}/tickets/${encodeURIComponent(ticketPath)}`, {
    method: 'DELETE',
  })
  if (!res.ok) {
    throw new Error('Failed to archive ticket')
  }
}

/** Batch assign tickets to new sessions. */
export async function assignTickets(boardId, tickets, { parallel = true } = {}) {
  const res = await workspaceFetch(`/boards/${boardId}/assign`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ tickets, parallel }),
  })
  if (!res.ok) {
    throw new Error('Failed to assign tickets')
  }
  return res.json()
}

/** Create a new swimlane. */
export async function createSwimlane(boardId, name) {
  const res = await workspaceFetch(`/boards/${boardId}/swimlanes`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  })
  if (!res.ok) {
    throw new Error('Failed to create swimlane')
  }
  return res.json()
}

/** Rename an existing swimlane. */
export async function renameSwimlane(boardId, swimlaneId, name) {
  const res = await workspaceFetch(`/boards/${boardId}/swimlanes/${swimlaneId}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  })
  if (!res.ok) {
    throw new Error('Failed to rename swimlane')
  }
  return res.json()
}

/** Delete a swimlane. Tickets in it become unsorted. */
export async function deleteSwimlane(boardId, swimlaneId) {
  const res = await workspaceFetch(`/boards/${boardId}/swimlanes/${swimlaneId}`, {
    method: 'DELETE',
  })
  if (!res.ok) {
    throw new Error('Failed to delete swimlane')
  }
}

/** Rename a column/state's display label. Folder and ID stay unchanged. */
export async function renameState(boardId, stateId, label) {
  const res = await workspaceFetch(`/boards/${boardId}/states/${stateId}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ label }),
  })
  if (!res.ok) {
    throw new Error('Failed to rename state')
  }
  return res.json()
}

/** Reorder columns/states to match the given key order. */
export async function reorderStates(boardId, keys) {
  const res = await workspaceFetch(`/boards/${boardId}/states/reorder`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ keys }),
  })
  if (!res.ok) {
    throw new Error('Failed to reorder states')
  }
  return res.json()
}

/** Reorder swimlanes to match the given ID order. */
export async function reorderSwimlanes(boardId, ids) {
  const res = await workspaceFetch(`/boards/${boardId}/swimlanes/reorder`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ ids }),
  })
  if (!res.ok) {
    throw new Error('Failed to reorder swimlanes')
  }
  return res.json()
}
